### PR TITLE
Logging Bridge For ZeroLog

### PR DESCRIPTION
**Pull request by AkhigbeEromo**
part of #5405 

- Worked on the [zerolog.Hook](https://pkg.go.dev/github.com/rs/zerolog#Hook) log bridge.
- Wrote Some tests for the functions

### CODE REVIEW

File: bridges/otelzerolog/hook.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
+package otelzerolog
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:1]: This is missing a copyright header.

There seem to be other lint issues with this PR. Can you run the `make precommit` hook and update?

Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook.go:1]: Okay will do just that

**Comment by AkhigbeEromo** [comment]
Hello @MrAlias, i have made the corrections. Is there anything you would like me to add

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add log support for the autoexport package. (#5733)
 - Add support for disabling the old runtime metrics using the `OTEL_GO_X_DEPRECATED_RUNTIME_METRICS=false` environment variable. (#5747)
 
+
```
Comment (MrAlias) [inline CHANGELOG.md:19]:
```suggestion
```

File: CHANGELOG.md
Comment (MrAlias) [inline CHANGELOG.md:1]: A changelog entry is required.

File: bridges/otelzerolog/go.mod [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module go.opentelemetry.io/contrib/bridges/otelzerolog
+
+go 1.22.1
```
Comment (MrAlias) [inline bridges/otelzerolog/go.mod:3]: This repo currently needs to support Go 1.21.

File: bridges/otelzerolog/hook.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+// If [WithLoggerProvider] is not provided, the returned SeverityHook will use the
+// global LoggerProvider.
+func NewSeverityHook(name string, options ...Option) *SeverityHook {
+	cfg := newConfig(options)
+	return &SeverityHook{
+		logger: cfg.logger(name),
+	}
+}
+
+// // Hook is a [zerolog.Hook] that sends all logging records it receives to
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:125]:
```suggestion
// SeverityHook is a [zerolog.Hook] that sends all logging records it receives to
```

File: bridges/otelzerolog/hook.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+// OpenTelemetry. See package documentation for how conversions are made.
+type SeverityHook struct {
+	logger log.Logger
+	levels zerolog.Level
+}
+
+// Levels returns the list of log levels we want to be sent to OpenTelemetry.
+func (h *SeverityHook) Levels() zerolog.Level {
+	return h.levels
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:134]: This is never set, it will always be empty.

File: bridges/otelzerolog/hook.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// // Hook is a [zerolog.Hook] that sends all logging records it receives to
+// OpenTelemetry. See package documentation for how conversions are made.
+type SeverityHook struct {
+	logger log.Logger
+	levels zerolog.Level
+}
+
+// Levels returns the list of log levels we want to be sent to OpenTelemetry.
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:132]: This doesn't return a list. Is it supposed to return the minimum logged severity?

File: bridges/otelzerolog/hook.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+func (h *SeverityHook) Levels() zerolog.Level {
+	return h.levels
+}
+
+// Run handles the passed record, and sends it to OpenTelemetry.
+func (h SeverityHook) Run(e *zerolog.Event, level zerolog.Level, msg string) error {
+	if level != zerolog.NoLevel {
+		e.Str("severity", level.String())
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:140]: Why is the severity being added as an attribute? This does not seem compliant with semantic converntions.

File: bridges/otelzerolog/hook.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+	}
+	h.logger.Emit(e.GetCtx(), h.convertEvent(e, level, msg))
+	return nil
+}
+
+func (h *SeverityHook) convertEvent(e *zerolog.Event, level zerolog.Level, msg string) log.Record {
+	var record log.Record
+	record.SetTimestamp(zerolog.TimestampFunc())
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:148]: The observed timestamp should also be set here.

File: bridges/otelzerolog/hook.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+	record.SetBody(log.StringValue(msg))
+	const sevOffset = zerolog.Level(log.SeverityDebug) - zerolog.DebugLevel
+	record.SetSeverity(log.Severity(level + sevOffset))
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:151]: This assumes the same scaling factor. OTel levels to not scale at this rate: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber

File: bridges/otelzerolog/hook.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+	fields := extractFields(e)
+
+	record.AddAttributes(convertFields(fields, msg)...)
+	return record
+}
+
+func extractFields(_ *zerolog.Event) map[string]interface{} {
+	// Here you would implement the logic to extract fields from the zerolog event
+	// This might involve using reflection or zerolog internals if necessary
+	fields := make(map[string]interface{})
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:161]: This is ... worrying, it is allocating a map just to make a conversion. Can the zero log fields not be directly converted sent to `SetAttribute` and skip this allocation?

File: bridges/otelzerolog/hook.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+	kvs := make([]log.KeyValue, 0, len(fields))
+	kvs = append(kvs, log.String("message", msg))
+	for k, v := range fields {
+		kvs = append(kvs, convertAttribute(k, v))
+	}
+	return kvs
+}
+
+func convertAttribute(key string, value interface{}) log.KeyValue {
+	switch v := value.(type) {
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:177]: Missing cases:

- map
- slice

File: bridges/otelzerolog/hook.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+	// Here you would implement the logic to extract fields from the zerolog event
+	// This might involve using reflection or zerolog internals if necessary
+	fields := make(map[string]interface{})
+	// Dummy implementation - replace with actual field extraction
+
+	return fields
+}
+
+func convertFields(fields map[string]interface{}, msg string) []log.KeyValue {
+	kvs := make([]log.KeyValue, 0, len(fields))
```
Comment (MrAlias) [inline bridges/otelzerolog/hook.go:168]: Similar here, this is making an allocation that is going to be potentially thrown away every time a message is emitted.

File: bridges/otelzerolog/hook_test.go
Comment (MrAlias) [inline bridges/otelzerolog/hook_test.go:1]: There is a fair amount of test coverage missing. If we plan to implement a method or function in the PR it should be accompanied with comprehensive tests.

If you need to break this into smaller PRs, adding a method/function stub that you will implement and test in a follow-up PR is acceptable. You will need to track that as an issue though and keep the bridge unreleased until all parts are complete.

**Review by MrAlias — COMMENTED** [review]
`versions.yaml` needs to be updated

File: bridges/otelzerolog/hook_test.go
Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook_test.go:1]: Thank you @MrAlias 
I will do just that

Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook_test.go:1]: I wanted to check in regarding this PR. While I haven't finished all the implementations yet, are the functions I've initialized sufficient, or do I need to add any additional functions? @MrAlias

Comment (MrAlias) [reply bridges/otelzerolog/hook_test.go:1]: I am more asking about adding test coverage for the functions that are included in this PR. There is a considerable amount of functionality being introduced here that is not tested.

File: bridges/otelzerolog/hook.go (hunk 8 above)
Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook.go:163]: I have added it as time.now(), hope that is okay

File: bridges/otelzerolog/hook.go (hunk 6 above)
Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook.go:142]: It now returns a list

File: bridges/otelzerolog/hook.go (hunk 5 above)
Comment (AkhigbeEromo) [reply bridges/otelzerolog/hook.go:144]: I just set it in the Severity Struct

File: bridges/otelzerolog/go.mod (hunk 3 above)
Comment (AkhigbeEromo) [reply bridges/otelzerolog/go.mod:3]: I resolved this in #5782

**Pull request closed without merging**